File: project1/week10/Danielle/word_search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        for i in range(m):
            for j in range(n):
                if self.dfs(board, word, i, j, 0):
                    return True
        return False

    def dfs(self, board, word, x, y, pos):
        if len(word) == pos:
            return True
        # boundary condition
        if x<0 or y<0 or x>=len(board) or y>=len(board[0]) or board[x][y]!=word[pos]:
            return False
        # search
        directions = [(-1,0),(1,0),(0,-1),(0,1)]
        temp = board[x][y]
        board[x][y] = -1
        for dx, dy in directions:
            if self.dfs(board, word, x+dx, y+dy, pos+1): 
                return True
        board[x][y] = temp
        return False

File: project2/week2/eunchong/wordSearch.cc
bool dfs(vector<vector<char>>& board, string word, int i, int j, int k) {
    // check if current coordinates are out of grid or the current cell doesn't match the current character of the word
    if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != word[k]) {
        return false;
    }
    // check if we have reached the end of the word
    if (k == word.length() - 1) {
        return true;
    }
    // mark the current cell as visited by replacing it with '/'
    char tmp = board[i][j];
    board[i][j] = '/';
    // check all 4 adjacent cells recursively
    bool res = dfs(board, word, i+1, j, k+1) ||
              dfs(board, word, i-1, j, k+1) ||
              dfs(board, word, i, j+1, k+1) ||
              dfs(board, word, i, j-1, k+1);
    // backtrack by replacing the current cell with its original value
    board[i][j] = tmp;
    return res;
}

bool exist(vector<vector<char>>& board, string word) {
    for (int i = 0; i < board.size(); i++) {
        for (int j = 0; j < board[0].size(); j++) {
            // start the search from every cell
            if (dfs(board, word, i, j, 0)) {
                return true;
            }
        }
    }
    return false;
}

/*
Time Complexity: O(4^N)
Space Complexity: O(N)
*/
